README.md
**NOTE:** All 0.x releases are kind of "alpha-versions" without expectations of interface backward-compatibility.

## Table of Contents
[mdz_unicode Overview](#mdz_unicode-Overview)<br>
[mdz_unicode Advantages](#mdz_unicode-Advantages)<br>
[mdz_unicode Usage](#mdz_unicode-Usage)<br>
[mdz_wchar Overview](#mdz_wchar-Overview)<br>
[mdz_utf8 Overview](#mdz_utf8-Overview)<br>
[mdz_utf16 Overview](#mdz_utf16-Overview)<br>
[mdz_utf32 Overview](#mdz_utf32-Overview)<br>
[Licensing info](#Licensing-info)<br>
[Credits](#Credits)

## mdz_unicode Overview

Wiki: [mdz_unicode Wiki]<br>
file: *"mdz_unicode.h"*

Please take a look at *"mdz_unicode.h"* file or [mdz_unicode Wiki] site for detailed functions descriptions.

[mdz_unicode Wiki]: https://github.com/maxdz-gmbh/mdz_unicode/wiki/mdz_unicode-overview

[mdz_unicode] - is a very lightweight, versatile and speedy C  library for handling Unicode strings, developed by [maxdz Software GmbH].The library supports UTF8, UTF16, UTF32, wchar strings. Source code of library is highly-portable, conforms to ANSI C 89/90 Standard. Builds for Win32/Win64, Linux, FreeBSD, Android, macOS are available.

Only shared/dynamically loaded libraries (*.so* and *.dll* files with import libraries) are available for evaluation testing purposes. Static libraries are covered by our commercial licenses.

**Linux** binaries are built against Linux Kernel 2.6.18 - and thus should be compatible with Debian (from ver. 4.0), Ubuntu (from ver. 8.04), Fedora (from ver. 9)

**FreeBSD** binaries - may be used from FreeBSD ver. 7.0

**Win32** binaries are built using Visual Studio Platform Toolset "v90", thus are compatible with Windows versions from Windows 2000.<br>
**Win64** binaries are built using Visual Studio Platform Toolset "v100", thus are compatible with Windows versions from Windows XP.<br>

**Android** x86/armeabi-v7a binaries - may be used from Android API level 16 ("Jelly Bean" ver. 4.1.x)<br>
**Android** x86_64/arm64-v8a binaries - may be used from Android API level 21 ("Lollipop" ver. 5.0)

**macOS** binaries - x86_64, from *MacOS X v10.6.0*

[mdz_unicode]: https://github.com/maxdz-gmbh/mdz_unicode
[maxdz Software GmbH]: https://maxdz.com/

## mdz_unicode Advantages

**1. High portability:** the whole code conforms to ANSI C 89/90 Standard. Multithreading/asynchronous part is POSIX compatible (under UNIX/Linux).

**2. Little dependencies:** basically *mdz_unicode* functions are only dependend on standard C-library memory-management/access functions. Multithreading part is dependend on POSIX *pthreads* API (under UNIX/Linux) and old process control/synchronization API (from Windows 2000). It means you can use library in your code withouth any further dependencies except standard platform libraries/APIs.

**3. Extended error-checking:** all functions preserve internal error-code pointing the problem. It is possible to use strict error-checking (when all preserved error-codes should be *MDZ_ERROR_NONE*) or "relaxed"-checking - when only returned *mdz_false* will indicate error.

**4. Extended control:** containers do only explicit operations. It means for example, when "insert" function is called with auto-reservation flag set in *mdz_false* - it will return error if there is not enough capacity in container. No implicit reservations will be made.

**5. Attached usage:** strings should not necessarily use dynamically-allocated memory - which may be not available on your embedded system (or if malloc()/free() are forbidden to use in your safety-critical software). Just attach container/data to your statically-allocated memory and use all strings functionality.

**6. Cache-friendly:** it is possible to keep controlling and data parts together in memory using "embedded part".

**7. Unicode support:** UTF-8, UTF-16, UTF-32 are supported.

**8. wchar_t support:** also wchar_t strings are supported, with 2 and 4 bytes-large *wchar_t* characters.

**9. Endianness-aware containers:** utf16 and utf32 containers are endiannes-aware thus may be used to produce and manipulate strings with pre-defined endianness even if endianness of host differs.

**10. Unicode "surrogate-pairs" awareness:** 2-byte Unicode strings correctly process/distinguish "surrogate-pairs" as 1 Unicode symbol.

**11. Asynchronous execution:** *insert* functions can be executed asynchronously

## mdz_unicode Usage

**unicode** is implemented with strict input parameters checking. It means *mdz_false* or some other error indication will be returned if one or several input parameters are invalid - even if such an invalidity doesn't lead to inconsistence (for example adding or removing 0 items).<br>

**Test license generation:** - in order to get free test-license, please proceed to our Shop page [maxdz Shop] and register an account. After registration you will be able to obtain free 14-days test-licenses for our products using "Obtain for free" button. 
Test license data should be used in *mdz_unicode_init()* call for library initialization.

**NOTE:** All 0.x releases are kind of "beta-versions" and can be used 1) only with test-license (during test period of 14 days, with necessity to re-generate license for the next 14 days test period) and 2) without expectations of interface backward-compatibility.

Several usage-scenarios are possible:
- low-level - raw C interface, using *mdz_unicode.h*, *mdz_utf8.h*, *mdz_utf16.h*, etc C-header files
- higher-level - using *MdzUnicode*, *MdzUtf8*, *MdzUtf16*, etc C++ "wrappers" around C-header files functions

[mdz_unicode Wiki]: https://github.com/maxdz-gmbh/mdz_unicode/wiki/mdz_unicode-overview
[maxdz Shop]: https://maxdz.com/shop.php

#### Code Example (low-level use)

*mdz_unicode_init()* with license information should be called for library initialization before any subsequent calls:

```
#include <mdz_unicode.h>

int main(int argc, char* argv[])
{
  /* mdz_unicode library initialization using test info retrieved after license generation (see "Test license generation" above) */
  
  mdz_bool bRet = mdz_unicode_init("<first-name-hash>", "<last-name-hash>", "<email-hash>", "<license-hash>");
  ...

  mdz_ansi_uninit(); /* call for un-initialization of library */
  
  return 0;
  
}
```

[mdz_utf8_create]: https://github.com/maxdz-gmbh/mdz_unicode/wiki/mdz_utf8_create
[mdz_utf8_destroy]: https://github.com/maxdz-gmbh/mdz_unicode/wiki/mdz_utf8_destroy

After library initialization call *[mdz_utf8_create]*() for **utf8** string creation. There should be also symmetric *[mdz_utf8_destroy]*() call for every create, otherwise allocated for string memory remains occupied:

```
#include <mdz_unicode.h>
#include <mdz_utf8.h>

int main(int argc, char* argv[])
{
  mdz_bool bRet = mdz_unicode_init("<first-name-hash>", "<last-name-hash>", "<email-hash>", "<license-hash>");
  
  // initialize pAnsi
  
  mdz_Utf8* pUtf8 = mdz_utf8_create(0); // create utf8-string
  ...
  ...
  // use pUtf8
  ...
  ...
  // destroy pUtf8
  
  mdz_utf8_destroy(&pUtf8); // after this pUtf8 should be NULL
  
  mdz_ansi_uninit();
  ...
}
```

Use *mdz_Utf8** pointer for subsequent library calls:

```
#include <mdz_unicode.h>
#include <mdz_utf8.h>

int main(int argc, char* argv[])
{
  mdz_bool bRet = mdz_unicode_init("<first-name-hash>", "<last-name-hash>", "<email-hash>", "<license-hash>");
  
  mdz_Utf8* pUtf8 = mdz_utf8_create(0); // create utf8-string

  // reserve memory for 5 elements
  
  bRet = mdz_utf8_reserve(pUtf8, 5);
  
  // insert 'b' in front position with auto-reservation if necessary
  
  bRet = mdz_utf8_insertAnsi(pUtf8, 0, "b", 1, mdz_true); // "b" after this call
  
  // append string with "cde" with auto-reservation if necessary
  
  bRet = mdz_utf8_insert(pUtf8, (size_t) -1, "cde", 3, mdz_true); // "bcde" after this call
  
  ...
  
  mdz_utf8_destroy(&pUtf8);
  
  mdz_ansi_uninit();
  ...
}
```

## mdz_wchar Overview
Wiki: [mdz_wchar Wiki]<br>
file: *"mdz_wchar.h"*

Please take a look at *"mdz_wchar.h"* file or [mdz_wchar Wiki] site for detailed functions descriptions.

[mdz_wchar Wiki]: https://github.com/maxdz-gmbh/mdz_containers/wiki/mdz_wchar-overview

## mdz_utf8 Overview
Wiki: [mdz_utf8 Wiki]<br>
file: *"mdz_utf8.h"*

Please take a look at *"mdz_utf8.h"* file or [mdz_utf8 Wiki] site for detailed functions descriptions.

[mdz_utf8 Wiki]: https://github.com/maxdz-gmbh/mdz_containers/wiki/mdz_utf8-overview

## mdz_utf16 Overview
Wiki: [mdz_utf16 Wiki]<br>
file: *"mdz_utf16.h"*

Please take a look at *"mdz_utf16.h"* file or [mdz_utf16 Wiki] site for detailed functions descriptions.

[mdz_utf16 Wiki]: https://github.com/maxdz-gmbh/mdz_containers/wiki/mdz_utf16-overview

## mdz_utf32 Overview
Wiki: [mdz_utf32 Wiki]<br>
file: *"mdz_utf32.h"*

Please take a look at *"mdz_utf32.h"* file or [mdz_utf32 Wiki] site for detailed functions descriptions.

[mdz_utf32 Wiki]: https://github.com/maxdz-gmbh/mdz_containers/wiki/mdz_utf32-overview

## Licensing info

Use of **mdz_unicode** library is regulated by license agreement in *LICENSE.txt*

Basically private non-commercial "test" usage is unrestricted. Commercial usage of library (incl. its source code) will be regulated by according license agreement.


HISTORY.txt
05.04.2021 (mon): Release 0.4
-----------------------------
- fixed handling of overlapping data and items
- minor changes in asynchronous execution

05.03.2021 (fri): Release 0.3
-----------------------------
- fixed handling of 4-byte wchar_t items in:
mdz_utf16_insertWchar_async
mdz_utf32_insertWchar_async
mdz_utf8_insertWchar_async
mdz_wchar_insertWchar_async

- added mdz_wchar_sizeof()

01.03.2021 (mon): Release 0.2
-----------------------------
- correction in "attach" functions

01.02.2020 (mon): Release 0.1
-------------------------------
Added functions:
- mdz_unicode_init
- mdz_unicode_init_attached
- mdz_unicode_uninit

- mdz_wchar_create
- mdz_wchar_create_attached
- mdz_wchar_destroy
- mdz_wchar_clear
- mdz_wchar_attachData
- mdz_wchar_reserve
- mdz_wchar_capacity
- mdz_wchar_size
- mdz_wchar_length
- mdz_wchar_offsetFromStart
- mdz_wchar_isAttachedData
- mdz_wchar_embedSize
- mdz_wchar_insertWchar_async
- mdz_wchar_insertWchar_string_async
- mdz_wchar_insertAnsi_async
- mdz_wchar_insertAnsi_string_async
- mdz_wchar_insertUtf8_async
- mdz_wchar_insertUtf8_string_async
- mdz_wchar_insertUtf16_async
- mdz_wchar_insertUtf16_string_async
- mdz_wchar_insertUtf32_async
- mdz_wchar_insertUtf32_string_async

- mdz_utf8_create
- mdz_utf8_create_attached
- mdz_utf8_destroy
- mdz_utf8_clear
- mdz_utf8_attachData
- mdz_utf8_reserve
- mdz_utf8_capacity
- mdz_utf8_size
- mdz_utf8_length
- mdz_utf8_offsetFromStart
- mdz_utf8_isAttachedData
- mdz_utf8_embedSize
- mdz_utf8_insertUtf8_async
- mdz_utf8_insertUtf8_string_async
- mdz_utf8_insertAnsi_async
- mdz_utf8_insertAnsi_string_async
- mdz_utf8_insertWchar_async
- mdz_utf8_insertWchar_string_async
- mdz_utf8_insertUtf16_async
- mdz_utf8_insertUtf16_string_async
- mdz_utf8_insertUtf32_async
- mdz_utf8_insertUtf32_string_async

- mdz_utf16_create
- mdz_utf16_create_attached
- mdz_utf16_destroy
- mdz_utf16_clear
- mdz_utf16_attachData
- mdz_utf16_reserve
- mdz_utf16_capacity
- mdz_utf16_size
- mdz_utf16_length
- mdz_utf16_endianness
- mdz_utf16_offsetFromStart
- mdz_utf16_isAttachedData
- mdz_utf16_embedSize
- mdz_utf16_insertUtf16_async
- mdz_utf16_insertUtf16_string_async
- mdz_utf16_insertAnsi_async
- mdz_utf16_insertAnsi_string_async
- mdz_utf16_insertWchar_async
- mdz_utf16_insertWchar_string_async
- mdz_utf16_insertUtf8_async
- mdz_utf16_insertUtf8_string_async
- mdz_utf16_insertUtf32_async
- mdz_utf16_insertUtf32_string_async

- mdz_utf32_create
- mdz_utf32_create_attached
- mdz_utf32_destroy
- mdz_utf32_clear
- mdz_utf32_attachData
- mdz_utf32_reserve
- mdz_utf32_capacity
- mdz_utf32_size
- mdz_utf32_length
- mdz_utf32_endianness
- mdz_utf32_offsetFromStart
- mdz_utf32_isAttachedData
- mdz_utf32_embedSize
- mdz_utf32_insertUtf32_async
- mdz_utf32_insertUtf32_string_async
- mdz_utf32_insertAnsi_async
- mdz_utf32_insertAnsi_string_async
- mdz_utf32_insertWchar_async
- mdz_utf32_insertWchar_string_async
- mdz_utf32_insertUtf8_async
- mdz_utf32_insertUtf8_string_async
- mdz_utf32_insertUtf16_async
- mdz_utf32_insertUtf16_string_async


mdz_unicode.h
/**
* \ingroup mdz_unicode library
*
* \author maxdz Software GmbH
*
* \par license
* This file is subject to the terms and conditions defined in file 'LICENSE.txt', which is part of this source code package.
*
* \par description
* mdz_unicode initialization functions.
*
* \par portability
* Source code of library conforms to ANSI C 89/90 Standard.
*
* \version 0.4
*
* \date 2021-04
*
*/

#ifndef MDZ_UNICODE_H
#define MDZ_UNICODE_H

#include "mdz_types.h"

#ifdef __cplusplus
extern "C"
{
#endif

/**
 * \defgroup Init functions
 */

/**
 * Initializes unicode library. This function should be called before any other function of the library.
 * \param pFirstNameHash - user first name hash code
 * \param pLastNameHash - user last name hash code
 * \param pEmailHash - user e-mail hash code
 * \param pLicenseHash - license hash code
 * \return:
 * mdz_true - if the initialization has succeed, otherwise false
 */
mdz_bool mdz_unicode_init(const uint32_t* pFirstNameHash, const uint32_t* pLastNameHash, const uint32_t* pEmailHash, const uint32_t* pLicenseHash);

/**
 * Initializes unicode library. This function should be caled before any other function of the library.
 * Memory for license data starts at position pStart. Size of internal initialization structure is returned in pSize.
 * \param pFirstNameHash - user first name hash code
 * \param pLastNameHash - user last name hash code
 * \param pEmailHash - user e-mail hash code
 * \param pLicenseHash - license hash code
 * \param pStart - memory start position of license data
 * \param nAreaSize - size of available memory from pStart in bytes. Should be large enough for license data (> 500 bytes)
 * \param pOutSize - actual size of placed license data in bytes
 * \return:
 * mdz_true - if the initialization has succeed, otherwise false
 */
mdz_bool mdz_unicode_init_attached(const uint32_t* pFirstNameHash, const uint32_t* pLastNameHash, const uint32_t* pEmailHash, const uint32_t* pLicenseHash, const char* pStart, size_t nAreaSize, size_t* pOutSize);

/**
 * Un-initializes unicode library and frees corresponding memory allocations.
 */
void mdz_unicode_uninit(void);

#ifdef __cplusplus
}
#endif

#endif


mdz_types.h
/**
 * \ingroup mdz libraries
 *
 * \author maxdz Software GmbH
 *
 * \par license
 * This file is subject to the terms and conditions defined in file 'LICENSE.txt', which is part of this source code package.
 *
 * \par description
 * Public types of mdz_containers library
 *
 * \par info
 * See additional info on mdz_containers library like version, portability, etc in mdz_containers.h
 */

#ifndef MDZ_TYPES_H
#define MDZ_TYPES_H

#include <stddef.h>

#include <stdint.h>
#include <limits.h>

#ifndef _WIN32
#include <pthread.h>
#endif

/**
 * Boolean type of mdz_containers library.
 */
#define mdz_false 0
#define mdz_true 1
typedef unsigned char mdz_bool;

#ifdef _WIN32
typedef void* HANDLE;
#endif

/**
 * Shared data for asynchronous calls
 */
struct mdz_asyncData
{
  /**
   * Pointer to mdz-container instance
   */
  void* m_pContainer;

  /**
   * Result of call. Invalid if call is not completely finished (m_bFinished is mdz_false)
   */
  size_t m_nResult;

  /**
   * Additional data returned by call (if any). Invalid if call is not completely finished (m_bFinished is mdz_false)
   */
  void* m_pData;

  /**
   * Thread on which the call is executed. May be used by client for wait operations
   */
#ifdef _WIN32
  HANDLE m_hThread;
#else
  pthread_t m_hThread;
#endif

  /**
   * mdz_true if the call is completely finished. Otherwise mdz_false (if interrupted/cancelled)
   */
  mdz_bool m_bFinished;

  /**
   * Should be set in mdz_true during call execution, to cancel the call. Otherwise mdz_false
   */
  mdz_bool m_bCancel;
};

/**
 * Attachment type
 */
enum mdz_attach_type
{
  /**
   * Size is 0 after attachment
   */
  MDZ_ATTACH_ZEROSIZE = 0,

  /**
   * Size is Capacity minus 1 after attachment; 0-terminator is present
   */
  MDZ_ATTACH_SIZE_TERMINATOR = 1,

  /**
   * Size is Capacity after attachment; without 0-terminator
   */
  MDZ_ATTACH_SIZE_NO_TERMINATOR = 2
};

/**
 * Data endianness
 */
enum mdz_endianness
{
  /**
   * Undefined endianness
   */
  MDZ_ENDIAN_UNDEFINED = 0,

  /**
   * Little-endian
   */
  MDZ_ENDIAN_LITTLE = 1,

  /**
   * Big-endian
   */
  MDZ_ENDIAN_BIG = 2,

  /**
   * Error endianness
   */
  MDZ_ENDIAN_ERROR = 3
};

/**
 * Result of comparison
 */
enum mdz_compare_result
{
  /**
   * Equal
   */
  MDZ_COMPARE_EQUAL = 0,

  /**
   * Non-equal
   */
  MDZ_COMPARE_NONEQUAL = 1,

  /**
   * Greater
   */
  MDZ_COMPARE_GREATER = 2,

  /**
   * Smaller
   */
  MDZ_COMPARE_SMALLER = 3,

  /**
   * Comparison ended with error
   */
  MDZ_COMPARE_ERROR = 4
};

/**
 * Method of find
 */
enum mdz_find_method
{
  /**
   * standard C-library implementation
   */
  MDZ_FIND_CLIB = 0,
  
  /**
   * Optimized Boyer-Moore-Horspool search
   */
  MDZ_FIND_BMH = 1,

  /**
   * Improved "brute-force" search
   */
  MDZ_FIND_BRUTE = 2,

  /**
   * Optimized search for monotone sub-string in monotone string
   */
  MDZ_FIND_MONOTONE = 3
};

enum mdz_error
{
  /**
   * No error
   */
  MDZ_ERROR_NONE = 0,

  /**
   * Invalid "data" parameter
   */
  MDZ_ERROR_DATA = 1,

  /**
   * Not enough capacity
   */
  MDZ_ERROR_CAPACITY = 2,

  /**
   * Invalid "offset" parameter
   */
  MDZ_ERROR_OFFSET = 3,

  /**
   * "Count" parameter is zero
   */
  MDZ_ERROR_ZEROCOUNT = 4,

  /**
   * "Count" parameter is too big
   */
  MDZ_ERROR_BIGCOUNT = 5,

  /**
   * "Left" parameter is too big
   */
  MDZ_ERROR_BIGLEFT = 6,

  /**
   * "Right" parameter is too big
   */
  MDZ_ERROR_BIGRIGHT = 7,

  /**
   * Invalid "items" parameter
   */
  MDZ_ERROR_ITEMS = 8,

  /**
   * Container is empty
   */
  MDZ_ERROR_EMPTY = 9,

  /**
   * Container is not empty
   */
  MDZ_ERROR_NONEMPTY = 10,

  /**
   * Invalid "sub-container" parameter
   */
  MDZ_ERROR_SUBCONTAINER = 11,

  /**
   * Container is attached
   */
  MDZ_ERROR_ATTACHED = 12,

  /**
   * Memory allocation error
   */
  MDZ_ERROR_ALLOCATION = 13,

  /**
   * Data content error
   */
  MDZ_ERROR_CONTENT = 14,

  /**
   * Invalid endiannes type
   */
  MDZ_ERROR_ENDIANNESS = 15,

  /**
   * Invalid attach type
   */
  MDZ_ERROR_ATTACH_TYPE = 16,

  /**
   * Invalid find method
   */
  MDZ_ERROR_FIND_METHOD = 17,

  /**
   * Memory allocation for thread error
   */
  MDZ_ERROR_THREAD_ALLOC = 18,

  /**
   * Thread start error
   */
  MDZ_ERROR_THREAD_START = 19,

  /**
   * Invalid "source string" parameter
   */
  MDZ_ERROR_SOURCE = 20,

  /**
   * Invalid attach terminator
   */
  MDZ_ERROR_ATTACH_TERMINATOR = 21,

  /**
   * Invalid wchar_t size (expected is 2 or 4 bytes)
   */
  MDZ_ERROR_WCHAR_SIZE = 22,

  /**
   * Overlappend areas error
   */
  MDZ_ERROR_OVERLAP = 23
};

#endif
